app/controllers/CyaMaintainProfileController.scala

```diff
@@ -191,26 +202,72 @@ class CyaMaintainProfileController @Inject() (
               NiphlNumberSummary.rowUpdate(request.userAnswers)
             ).flatten
           )
-          Future.successful(Ok(view(list, routes.CyaMaintainProfileController.onSubmitNiphlNumber)))
+          Future.successful(
+            Ok(view(list, routes.CyaMaintainProfileController.onSubmitNiphlNumber))
+              .removingFromSession(dataUpdated, pageUpdated, dataRemoved, dataAdded)
+          )
         case Left(errors) =>
           Future.successful(logErrorsAndContinue(errorMessage, routes.ProfileController.onPageLoad(), errors))
       }
     }
 
   def onSubmitNiphlNumber(): Action[AnyContent] = (identify andThen getData andThen requireData).async {
     implicit request =>
-      TraderProfile.validateNiphlsUpdate(request.userAnswers) match {
-        case Right(niphlNumber) =>
-          traderProfileConnector.getTraderProfile(request.eori).flatMap { traderProfile =>
-            val updatedProfile = traderProfile.copy(niphlNumber = niphlNumber)
-            auditService.auditMaintainProfile(traderProfile, updatedProfile, request.affinityGroup)
-            for {
-              _ <- traderProfileConnector.submitTraderProfile(updatedProfile, request.eori)
-            } yield Redirect(navigator.nextPage(CyaMaintainProfilePage, NormalMode, request.userAnswers))
-          }
-        case Left(errors)       =>
-          Future.successful(logErrorsAndContinue(errorMessage, routes.ProfileController.onPageLoad(), errors))
+      (for {
+        niphlNumber      <- handleValidateError(TraderProfile.validateNiphlNumber(request.userAnswers))
+        oldTraderProfile <- traderProfileConnector.getTraderProfile(request.eori)
+        newTraderProfile <- Future.successful(oldTraderProfile.copy(niphlNumber = niphlNumber))
+        _                 = auditService.auditMaintainProfile(oldTraderProfile, newTraderProfile, request.affinityGroup)
+        _                <- submitTraderProfileIfValueChanged(newTraderProfile, oldTraderProfile, NiphlNumberUpdatePage, request.eori)
+      } yield Redirect(navigator.nextPage(CyaMaintainProfilePage, NormalMode, request.userAnswers))
+        .addingToSession(
+          dataUpdated -> isValueChanged(newTraderProfile, oldTraderProfile, NiphlNumberUpdatePage).toString
+        )
+        .addingToSession(dataAdded -> isValueAdded(newTraderProfile, oldTraderProfile, NiphlNumberUpdatePage).toString)
+        .addingToSession(pageUpdated -> niphlNumberUpdatePage)).recover { case e: TraderProfileBuildFailure =>
+        logErrorsAndContinue(e.getMessage, routes.ProfileController.onPageLoad())
       }
   }
 
+  private def isValueChanged(newTraderProfile: TraderProfile, oldTraderProfile: TraderProfile, page: Page): Boolean =
```

Comment:
below are a load of boolean checks to work out what's changed and then we store that on the end of each submit